File: src/priv/bajdcc/util/lexer/token/Token.java
package priv.bajdcc.util.lexer.token;

import java.math.BigDecimal;
import java.math.BigInteger;

import priv.bajdcc.util.Position;

/**
 * 单词
 * 
 * @author bajdcc
 */
public class Token implements Cloneable {
	/**
	 * 单词类型
	 */
	public TokenType kToken = TokenType.ERROR;

	/**
	 * 数据
	 */
	public Object object = null;

	/**
	 * 位置
	 */
	public Position position = new Position();

	public String toRealString() {
		return getRealString(kToken, object);
	}

	public static String getRealString(TokenType type, Object object) {
		if (object == null) {
			return "";
		}
		switch (type) {
		case KEYWORD:
			return ((KeywordType) object).getName();
		case OPERATOR:
			return ((OperatorType) object).getName();
		case STRING:
			return "\"" + object.toString() + "\"";
		case CHARACTER:
			char ch = (char) object;
			return "'"
					+ (Character.isISOControl(ch) ? String.format("\\u%04x",
							(int) ch) : ch + "") + "'";
		case BOOL:
		case DECIMAL:
		case INTEGER:
		case ID:
			return object.toString();
		default:
			return "";
		}
	}

	private String toSimpleString() {
		switch (kToken) {
		case KEYWORD:
			return ((KeywordType) object).getName();
		case OPERATOR:
			return ((OperatorType) object).getName();
		default:
			return "";
		}
	}

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();
		sb.append(String.format("%04d,%03d:\t%s\t%s %s", position.iLine,
				position.iColumn, kToken.getName(), object == null ? "(null)"
						: object.toString(), toSimpleString()));
		return sb.toString();
	}

	public Token copy() {
		try {
			return (Token) clone();
		} catch (CloneNotSupportedException e) {
			e.printStackTrace();
		}
		return null;
	}

	public static Token createFromObject(Object object) {
		Token token = new Token();
		if (object instanceof String) {
			token.kToken = TokenType.STRING;
		} else if (object instanceof Character) {
			token.kToken = TokenType.CHARACTER;
		} else if (object instanceof BigInteger) {
			token.kToken = TokenType.INTEGER;
		} else if (object instanceof BigDecimal) {
			token.kToken = TokenType.DECIMAL;
		} else if (object instanceof Boolean) {
			token.kToken = TokenType.BOOL;
		} else {
			token.kToken = TokenType.ERROR;
			return token;
		}
		token.object = object;
		return token;
	}
}

File: src/main/java/com/bajdcc/util/lexer/token/Token.kt
package com.bajdcc.util.lexer.token

import com.bajdcc.util.Position

/**
 * 单词
 *
 * @author bajdcc
 */
class Token : Cloneable {
    /**
     * 单词类型
     */
    var type = TokenType.ERROR

    /**
     * 数据
     */
    var obj: Any? = null

    /**
     * 位置
     */
    var position = Position()

    fun toRealString(): String {
        return getRealString(type, obj ?: "")
    }

    private fun toSimpleString(): String {
        return when (type) {
            TokenType.KEYWORD -> (obj as KeywordType).desc
            TokenType.OPERATOR -> (obj as OperatorType).desc
            else -> ""
        }
    }

    override fun toString(): String {
        return String.format("%04d,%03d:\t%s\t%s %s", position.line,
                position.column, type.desc,
                obj?.toString() ?: "(null)",
                if (obj == null) "(null)" else toSimpleString())
    }

    fun copy(): Token {
        return clone() as Token
    }

    companion object {
        fun getRealString(type: TokenType, obj: Any): String = when (type) {
            TokenType.KEYWORD -> (obj as KeywordType).desc
            TokenType.OPERATOR -> (obj as OperatorType).desc
            TokenType.STRING -> "\"" + obj.toString() + "\""
            TokenType.CHARACTER -> {
                val ch = obj as Char
                String.format("'%s'",
                        (if (Character.isISOControl(ch))
                            String.format("\\u%04x", ch.toInt())
                        else
                            ch + ""))
            }
            TokenType.BOOL, TokenType.DECIMAL, TokenType.INTEGER, TokenType.ID -> obj.toString()
            else -> ""
        }

        fun createFromObject(obj: Any): Token {
            val token = Token()
            when (obj) {
                is String -> token.type = TokenType.STRING
                is Char -> token.type = TokenType.CHARACTER
                is Long -> token.type = TokenType.INTEGER
                is Double -> token.type = TokenType.DECIMAL
                is Boolean -> token.type = TokenType.BOOL
                is Int -> token.type = TokenType.POINTER
                else -> {
                    token.type = TokenType.ERROR
                    return token
                }
            }
            token.obj = obj
            return token
        }
    }
}